lib/nrf_inbuilt_key/nrf_inbuilt_key.c

```diff
@@ -0,0 +1,341 @@
+	}
+
+	if (retval == 0) {
+		struct at_param_list cmng_list;
+
+		at_params_list_init(&cmng_list, 5);
+		at_parser_params_from_str(scratch_buf, NULL, &cmng_list);
+
+		size_t len;
+
+		at_params_size_get(&cmng_list, 4, &len);
+		if (len < *p_buffer) {
+			at_params_list_free(&cmng_list);
+			return -ENOBUFS;
+		}
+		at_params_string_get(&cmng_list, 4, p_buffer, &len);
+		*p_buffer_len = len;
+		at_params_list_free(&cmng_list);
+	}
+
+	return retval;
+}
+
+int nrf_inbuilt_key_delete(nrf_sec_tag_t sec_tag,
+			   nrf_key_mgnt_cred_type_t cred_type)
+{
+	u16_t scratch_buf_len = sizeof(NRF_INBUILT_KEY_OP_RM) +
+				(NRF_INBUILT_KEY_CHAR_SIZE_MAX_INT32 * 2) +
+				(NRF_INBUILT_KEY_COMMA_SIZE * 2) +
+				NRF_INBUILT_KEY_CR_CHAR_SIZE +
+				NRF_INBUILT_KEY_LF_CHAR_SIZE +
+				NRF_INBUILT_KEY_ZERO_TERMINATOR_SIZE;
+
+	u8_t *p_cmd = &scratch_buf[0];
+
+	int exp_len = snprintf((char *)p_cmd, scratch_buf_len, "%s,%u,%u\r\n",
+			       NRF_INBUILT_KEY_OP_RM, (u32_t)sec_tag,
+			       (u8_t)cred_type);
+
+	if (exp_len >= scratch_buf_len) {
+		return -ENOBUFS;
+	}
+
+	bool cmee_activated = false;
+	int cmee_active_result = cmee_active();
+
+	if (cmee_active_result < 0) {
+		return -EFAULT;
+	}
+
+	if (cmee_active_result == 0) {
+		cmee_set(true);
+		cmee_activated = true;
+	}
+
+	enum at_cmd_state state;
+	int result =
+		at_cmd_write(scratch_buf, scratch_buf, scratch_buf_len, &state);
+	if (cmee_activated) {
+		cmee_set(false);
+	}
+
+	int retval = 0;
+
+	if ((result != 0) && (state == AT_CMD_ERROR_CME)) {
+		retval = translate_error(result);
+	}
+
+	return retval;
+}
+
+int nrf_inbuilt_key_permission_set(nrf_sec_tag_t sec_tag,
+				   nrf_key_mgnt_cred_type_t cred_type,
+				   u8_t perm_flags)
+{
+	int retval = -EOPNOTSUPP;
+	return retval;
+}
+
+int nrf_inbuilt_key_exists(nrf_sec_tag_t sec_tag,
+			   nrf_key_mgnt_cred_type_t cred_type, bool *p_exists,
+			   u8_t *p_perm_flags)
+{
+	u16_t scratch_buf_len = sizeof(NRF_INBUILT_KEY_OP_LS) +
+				(NRF_INBUILT_KEY_CHAR_SIZE_MAX_INT32 * 2) +
+				(NRF_INBUILT_KEY_COMMA_SIZE * 2) +
+				NRF_INBUILT_KEY_CR_CHAR_SIZE +
+				NRF_INBUILT_KEY_LF_CHAR_SIZE +
+				NRF_INBUILT_KEY_ZERO_TERMINATOR_SIZE;
+
+	u8_t *p_cmd = &scratch_buf[0];
+
+	int exp_len = snprintf((char *)p_cmd, scratch_buf_len, "%s,%u,%u\r\n",
+			       NRF_INBUILT_KEY_OP_LS, (u32_t)sec_tag,
+			       (u8_t)cred_type);
+
+	if (exp_len >= scratch_buf_len) {
+		return -ENOBUFS;
+	}
+
+
+	bool cmee_activated = false;
+	int cmee_active_result = cmee_active();
+
+	if (cmee_active_result < 0) {
+		return -EFAULT;
+	}
+
+	if (cmee_active_result == 0) {
+		cmee_set(true);
+		cmee_activated = true;
+	}
+
+	enum at_cmd_state state;
+	int result = at_cmd_write(scratch_buf, scratch_buf, sizeof(scratch_buf),
+				  &state);
+	if (cmee_activated) {
+		cmee_set(false);
+	}
+
+	int retval = 0;
+
+	if ((result != 0) && (state == AT_CMD_ERROR_CME)) {
+		retval = translate_error(result);
+	}
+
+	if ((retval == 0) && (strlen(scratch_buf) > 0)) {
+		*p_exists = true;
```

Comment:
You should check that neither `p_exists` nor `p_perm_flags` are `NULL` at the beginning of the function.

Reply:
Good point.